src/main/java/org/jitsi/videobridge/xmpp/MediaStreamTrackFactory.java

```diff
@@ -383,157 +106,159 @@ private static void removeReferences(TrackSsrcs trackSsrcs, List<SourcePacketExt
+            for (SourceGroupPacketExtension sourceGroup : groupsCopy)
+            {
+                List<SourcePacketExtension> groupSources
+                    = sourceGroup.getSources();
 
-        public SecondarySsrc(long ssrc, String type)
-        {
-            this.ssrc = ssrc;
-            this.type = type;
-        }
-    }
+                if (groupSources.size() != 2)
+                {
+                    logger.warn(
+                            "A valid RTX or FEC group has exactly 2 ssrcs.");
+                    continue;
+                }
 
-    /**
-     * Groups a set of secondary ssrcs for a single primary ssrc (e.g. its
-     * rtx and fec ssrcs)
-     */
-    private static class SecondarySsrcs
-        implements Iterable<SecondarySsrc>
-    {
-        public List<SecondarySsrc> secondarySsrcs;
+                long primarySSRC = groupSources.get(0).getSSRC();
+    
+                MediaStreamTrackDesc track = addPlainTrack(primarySSRC);
 
-        public SecondarySsrcs(List<SecondarySsrc> secondarySsrcs)
+                // Set the secondary SSRC for the matching encodings.
+                long secondarySSRC = groupSources.get(1).getSSRC();
+                tracksMap.put(secondarySSRC, track);
+                for (RTPEncodingDesc encoding : track.getRTPEncodings())
+                {
+                    if (encoding.getPrimarySSRC() == primarySSRC)
+                    {
+                        encoding.addSecondarySsrc(
+                                secondarySSRC,
+                                semanticsMap.get(sourceGroup.getSemantics()));
+                    }
+                }
+            }
+	}
+	
+        // Process the free ssrcs
+        if (sources != null && !sources.isEmpty())
         {
-            this.secondarySsrcs = secondarySsrcs;
+            for (SourcePacketExtension source : sources)
+            {
+                long primarySSRC = source.getSSRC();
+                addPlainTrack(primarySSRC);
+            }
         }
 
-        @Override
-        public Iterator<SecondarySsrc> iterator()
-        {
-            return secondarySsrcs.iterator();
-        }
+        Collection<MediaStreamTrackDesc> tracks = tracksMap.values();
+	return tracks.toArray(new MediaStreamTrackDesc[tracks.size()]);
     }
 
-    /**
-     * Describes one or more primary video ssrcs for a single
-     * {@link MediaStreamTrackDesc}
-     */
-    private static class TrackSsrcs
-        implements Iterable<Long>
+    private MediaStreamTrackDesc addPlainTrack(long primarySSRC)
     {
-        private List<Long> trackSsrcs;
+        MediaStreamTrackDesc track = tracksMap.get(primarySSRC);
 
-        public TrackSsrcs(List<Long> trackSsrcs)
+        // Check if we already have a track for this SSRC.
+        if (track == null)
         {
-            this.trackSsrcs = trackSsrcs;
+            // It looks like this is the first time we see this SSRC,
+            // so we create a new track for it.
```

Comment:
i liked having it where an ssrc was removed from source groups/sources completely once it was processed rather than dealing with the possibility of adding the same track twice, any reason to go back to this?